packages/@lwc/engine-core/src/framework/stylesheet.ts

```diff
@@ -22,127 +21,87 @@ export type StylesheetFactory = (
+
+    // Reset the styling token applied to the host element.
     const oldHostAttribute = context.hostAttribute;
     if (!isUndefined(oldHostAttribute)) {
         renderer.removeAttribute(elm, oldHostAttribute);
     }
 
-    // Reset the scoping attributes associated to the context.
-    context.hostAttribute = context.shadowAttribute = undefined;
+    // Apply the new template styling token to the host element, if the new template has any
+    // associated stylesheets.
+    if (
+        !isUndefined(newStylesheetTokens) &&
+        !isUndefined(newStylesheets) &&
+        newStylesheets.length !== 0
+    ) {
+        newHostAttribute = newStylesheetTokens.hostAttribute;
+        newShadowAttribute = newStylesheetTokens.shadowAttribute;
+
+        renderer.setAttribute(elm, newHostAttribute, '');
+    }
+
+    // Update the styling tokens present on the context object.
+    context.hostAttribute = newHostAttribute;
+    context.shadowAttribute = newShadowAttribute;
 }
 
-/**
- * Apply/Update the styling token applied to the host element.
- */
-export function applyStyleAttributes(vm: VM, hostAttribute: string, shadowAttribute: string): void {
-    const { context, elm, renderer } = vm;
+export function getStylesheetsContent(vm: VM, template: Template): string[] {
+    const { stylesheets: factories, stylesheetTokens: tokens } = template;
+    const { syntheticShadow: useSyntheticShadow } = vm.renderer;
 
-    // Remove the style attribute currently applied to the host element.
-    renderer.setAttribute(elm, hostAttribute, '');
+    const stylesheets: string[] = [];
 
-    context.hostAttribute = hostAttribute;
-    context.shadowAttribute = shadowAttribute;
-}
+    if (!isUndefined(factories) && !isUndefined(tokens)) {
+        const hostSelector = useSyntheticShadow ? `[${tokens.hostAttribute}]` : '';
+        const shadowSelector = useSyntheticShadow ? `[${tokens.shadowAttribute}]` : '';
 
-function collectStylesheets(
-    stylesheets: StylesheetFactory[],
-    hostSelector: string,
-    shadowSelector: string,
-    isNative: boolean,
-    aggregatorFn: (content: string) => void
-): void {
-    forEach.call(stylesheets, (sheet) => {
-        if (isArray(sheet)) {
-            collectStylesheets(sheet, hostSelector, shadowSelector, isNative, aggregatorFn);
-        } else {
-            aggregatorFn(sheet(hostSelector, shadowSelector, isNative));
-        }
-    });
-}
+        for (let i = 0; i < factories.length; i++) {
+            const factory = factories[i];
 
-export function evaluateCSS(
-    vm: VM,
-    stylesheets: StylesheetFactory[],
-    hostAttribute: string,
-    shadowAttribute: string
-): VNode | null {
-    if (process.env.NODE_ENV !== 'production') {
-        assert.isTrue(isArray(stylesheets), `Invalid stylesheets.`);
+            if (isArray(factory)) {
+                for (let j = 0; j < factory.length; j++) {
+                    stylesheets.push(factory[j](hostSelector, shadowSelector, !useSyntheticShadow));
```

Comment:
I think this code is assuming that this is not recursive, which it is. Any import css at any level implies that you're going to have an array or arrays. Can you clarify?

Reply:
we probably don't have this for this though.

Reply:
>I think this code is assuming that this is not recursive, which it is. Any import css at any level implies that you're going to have an array or arrays.

@pmdartus, i think @caridy has a valid point, ex:
```css
/* c/A -> c/B -> c/C */
/* A */
@import 'c/B';

/* B */
@import 'c/C';

/* C */
p { color: red; }
```